File: python/2015/day25.py
# vi: set shiftwidth=4 tabstop=4 expandtab:
import datetime
import re

# To continue, please consult the code grid in the manual.  Enter the code at row 123, column 45.
code_re = re.compile(r"^To continue, please consult the code grid in the manual.  Enter the code at row (\d+), column (\d+).$")

def get_pos_from_line(string):
    return tuple(int(s) for s in code_re.match(string).groups())

def get_pos_from_file(file_path="../../resources/year2015_day25_input.txt"):
    with open(file_path) as f:
        for l in f:
            return get_pos_from_line(l.strip())

def get_index(pos):
    row, col = pos
    n = row + col - 2
    return col + n*(n+1) // 2

def get_code_at_index(n, start=20151125, mult=252533, mod=33554393):
    return (start * pow(mult, n-1, mod)) % mod

def get_code(pos):
    return get_code_at_index(get_index(pos))

def run_tests():
    assert get_index((4, 2)) == 12
    assert get_index((1, 5)) == 15
    res1 = 20151125
    res2 = 31916031
    res3 = 18749137
    res20 = 15514188
    assert get_code_at_index(1) == res1
    assert get_code((1, 1)) == res1
    assert get_code_at_index(2) == res2
    assert get_code((2, 1)) == res2
    assert get_code_at_index(3) == res3
    assert get_code((1, 2)) == res3
    assert get_code_at_index(20) == res20
    assert get_code((2, 5)) == res20

def get_solutions():
    print(get_code(get_pos_from_file()) == 2650453)


if __name__ == "__main__":
    begin = datetime.datetime.now()
    run_tests()
    get_solutions()
    end = datetime.datetime.now()
    print(end - begin)

File: rust/src/2015/day25.rs
// vi: set shiftwidth=4 tabstop=4 expandtab:
use common::input::check_answer;
use common::input::get_answers;
use common::input::get_first_line_from_file;
use std::time::Instant;

const INPUT_FILEPATH: &str = "../resources/year2015_day25_input.txt";
const ANSWERS_FILEPATH: &str = "../resources/year2015_day25_answer.txt";

type Int = u64;
type InputContent = (Int, Int);

const fn get_input_from_str(_string: &str) -> InputContent {
    // TODO: Parse string
    (2978, 3083)
}

const fn get_index(pos: (Int, Int)) -> Int {
    let (row, col) = pos;
    let n = row + col - 2;
    col + n * (n + 1) / 2
}

// From https://rob.co.bb/posts/2019-02-10-modular-exponentiation-in-rust/
const fn mod_pow(mut base: Int, mut exp: Int, modulus: Int) -> Int {
    if modulus == 1 {
        return 0;
    }
    let mut result = 1;
    base %= modulus;
    while exp > 0 {
        if exp % 2 == 1 {
            result = result * base % modulus;
        }
        exp >>= 1;
        base = base * base % modulus;
    }
    result
}

const fn get_code_at_index(n: Int, start: Int, mult: Int, modulus: Int) -> Int {
    (start * mod_pow(mult, n - 1, modulus)) % modulus
}

const fn part1(pos: InputContent) -> Int {
    get_code_at_index(get_index(pos), 20_151_125, 252_533, 33_554_393)
}

fn main() {
    let before = Instant::now();
    let data = get_input_from_str(&get_first_line_from_file(INPUT_FILEPATH));
    let (ans, _) = get_answers(ANSWERS_FILEPATH);
    let solved = true;
    let res = part1(data);
    check_answer(&res.to_string(), ans, solved);
    println!("Elapsed time: {:.2?}", before.elapsed());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_functions() {
        assert_eq!(get_index((4, 2)), 12);
        assert_eq!(get_index((1, 5)), 15);
        // TODO: Add more
    }
}
